File: sample/qp_active_set_demo_1.cpp
/**
 * @file qp_active_set_demo_1.cpp
 * @brief Demonstrates solving a quadratic programming (QP) problem using an
 * active set solver.
 *
 * This example sets up and solves a simple QP problem with two variables and
 * four constraints. The problem is defined by specifying the quadratic cost
 * matrix, linear term, constraint matrix, and constraint bounds. The active set
 * QP solver is used to find the optimal solution, which is then printed to the
 * standard output.
 */
#include <iostream>

#include "python_numpy.hpp"
#include "python_optimization.hpp"

using namespace PythonNumpy;
using namespace PythonOptimization;

int main(void) {
  /* Define Problem */
  constexpr std::size_t NUMBER_OF_VARIABLES = 2;
  constexpr std::size_t NUMBER_OF_CONSTRAINTS = 4;

  auto qp_solver = make_QP_ActiveSetSolver<double, NUMBER_OF_VARIABLES,
                                           NUMBER_OF_CONSTRAINTS>();

  auto E = make_DiagMatrixIdentity<double, NUMBER_OF_VARIABLES>();
  auto L = make_DenseMatrix<NUMBER_OF_VARIABLES, 1>(5.0, 8.0);

  auto M = make_DenseMatrix<NUMBER_OF_CONSTRAINTS, NUMBER_OF_VARIABLES>(
      1.0, 0.0, 0.0, 1.0, -1.0, 0.0, 0.0, -1.0);

  auto Gamma = make_DenseMatrix<NUMBER_OF_CONSTRAINTS, 1>(4.0, 6.0, 0.0, 0.0);

  /* Solve QP Problem */
  auto x_opt = qp_solver.solve(E, L, M, Gamma);

  /* Output Result */
  std::cout << "Optimal Solution:" << std::endl;
  for (std::size_t i = 0; i < NUMBER_OF_VARIABLES; ++i) {
    std::cout << "x[" << i << "] = " << x_opt(i, 0) << std::endl;
  }

  return 0;
}

File: sample/qp_active_set_demo_1.py
"""
This script demonstrates how to solve a quadratic programming (QP) problem using the QP_ActiveSetSolver class from the python_optimization.qp_active_set module.
It sets up a simple QP with two variables and four linear inequality constraints, then solves for the optimal solution using the active set method.
"""
import os

import sys
sys.path.append(os.getcwd())

import numpy as np

from python_optimization.qp_active_set import QP_ActiveSetSolver

# QP parameters
E = np.eye(2)
L = np.array([[5.0], [8.0]])
M = np.array([[1, 0],   # x <= 4
              [0, 1],   # y <= 6
              [-1, 0],   # x >= 0  (−x <= 0)
              [0, -1]])  # y >= 0  (−y <= 0)

gamma = np.array([[4.0],
                  [6.0],
                  [0.0],
                  [0.0]])

# Run solver
solver = QP_ActiveSetSolver(number_of_variables=E.shape[0],
                            number_of_constraints=M.shape[0]
                            )
x_opt = solver.solve(E, L, M, gamma)

print("Optimal solution: x =", x_opt)
